File: C++/Problems/Problems/problem_2537.cpp
#include "problem_2537.h"

#include <unordered_map>

bool problem_2537::test() {
    std::vector nums { 3, 1, 4, 3, 2, 2, 4 };
    int k = 2;

    int expected = 4;

    auto result = countGood(nums, k);

    std::cout << result << std::endl;

    return expected == result;
}

long long problem_2537::countGood(std::vector<int> &nums, int k) {
    int n = nums.size();
    std::unordered_map<int, int> counts;
    int l = 0;
    int r = -1;
    int p = 0;
    long long result = 0;
    while (r < n - 1)
    {
        while (p < k && r < n - 1)
        {
            r++;
            if (!counts.contains(nums[r]))
                counts[nums[r]] = 0;
            p += counts[nums[r]];
            counts[nums[r]]++;
        }
        while (p >= k)
        {
            result += n - r;
            counts[nums[l]]--;
            p -= counts[nums[l]];
            l++;
        }
    }

    return result;
}

File: Python/Problems/Problems/problem2537.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem2537(Problem):
    def test(self) -> bool:
        nums = [3, 1, 4, 3, 2, 2, 4]
        k = 2

        expected = 4

        result = self.countGood(nums, k)

        print(result)

        return expected == result

    def countGood(self, nums: List[int], k: int) -> int:
        n = len(nums)
        counts = dict()
        l = 0
        r = -1
        p = 0
        result = 0
        while r < n - 1:
            while p < k and r < n - 1:
                r += 1
                if nums[r] not in counts:
                    counts[nums[r]] = 0
                p += counts[nums[r]]
                counts[nums[r]] += 1
            while p >= k:
                result += n - r
                counts[nums[l]] -= 1
                p -= counts[nums[l]]
                l += 1

        return result